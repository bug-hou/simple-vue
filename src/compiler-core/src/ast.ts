export const enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPOUND_EXPRESSION
}

export const enum TagType {
  START,
  END
}