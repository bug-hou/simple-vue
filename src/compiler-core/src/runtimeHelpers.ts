export const TO_DISPLAY_STRING = Symbol("toDisplayString");
export const CREATE_ELEMENT_VNODE = Symbol("createElementVnode");

export const helperMapName = {
  [TO_DISPLAY_STRING]: "_toDisplayString",
  [CREATE_ELEMENT_VNODE]: "_createElementVnode"
}