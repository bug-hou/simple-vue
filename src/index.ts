export * from "./runtime-core";
export * from "./runtime-dom"