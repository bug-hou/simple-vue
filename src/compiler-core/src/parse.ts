import { NodeTypes, TagType } from "./ast";
import { AST, Context, ElementType, InterpolationType, TextType } from "./type/type";

export function baseParse(content: string): AST {
  const context = createParseContext(content);

  return createRoot(parseChildren(context))
}

/* 
解析根对象
*/
function createRoot(children: any[]): AST {
  return {
    type: NodeTypes.ROOT,
    children
  }
}


/* 
解析上下文
*/
function createParseContext(context: string): Context {
  return {
    source: context
  }
}

/* 
解析children
*/
function parseChildren(context: Context, tag?: string) {
  const nodes: any[] = [];
  while (isEnd(context, tag)) {
    let node;
    const s = context.source;
    if (s.startsWith("{{")) {
      node = parseInterpolation(context);
    } else if (s.startsWith("<")) {
      // 检测标签:<d
      if (/[a-z]/i.test(s[1])) {
        node = parseElement(context);
      }
    }

    if (!node) {
      node = parseText(context)
    }
    nodes.push(node)
  }
  return nodes;
}

function isEnd(context: Context, tag?: string) {
  const s = context.source;
  if (tag && s.startsWith(`</`)) {
    if (s.startsWith(`</${tag}`)) {
      return false;
    } else {
      // 防止<div><span></div>,语法错误
      throw new SyntaxError("缺少结束标签" + tag)
    }
  }
  return !(s.length === 0)
}

// 解析mustache函数
function parseInterpolation(context: Context): InterpolationType {
  const openDelimiter = "{{";
  const closeDelimiter = "}}";
  // {{message}}
  /* 
    先获取到}}对应的下标，在获取到message值
  */
  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length);
  // 将字符串向前推进2个单位
  context.source = advanceBy(context.source, openDelimiter.length);

  const rawContentLength = closeIndex - openDelimiter.length;

  const rawContent = parseTextDate(context.source, rawContentLength)
  // 去掉空格
  const content = rawContent.trim();

  // 在把字符串先前推进
  context.source = advanceBy(context.source, closeIndex)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content
    }
  }
}

// 推进字符串
function advanceBy(content: string, len: number) {
  return content.slice(len)
}

function parseElement(context: Context): ElementType {
  const element = paserTag(context, TagType.START);
  element.children = parseChildren(context, element.tag);
  paserTag(context, TagType.END);
  return element;
}

function paserTag(context: Context, type: any): ElementType | undefined {
  const match = /^<\/?([a-z]{1,})/i.exec(context.source);
  let tag: string;
  if (match) {
    tag = match[1];
    context.source = advanceBy(context.source, match[0].length + 1)
  }
  if (type === TagType.END) {
    return;
  }
  return {
    type: NodeTypes.ELEMENT,
    tag
  }
}

function parseText(context: Context): TextType {
  let endIndex = context.source.length;
  let endTokens = ["{{", "<"];

  endTokens.forEach(item => {
    let index = context.source.indexOf(item);
    if (index !== -1) {
      endIndex = endIndex > index ? index : endIndex;
    }
  })

  const s = parseTextDate(context.source, endIndex);
  context.source = advanceBy(context.source, s.length)
  return {
    type: NodeTypes.TEXT,
    content: s
  }
}

function parseTextDate(content: string, length: number) {
  return content.slice(0, length)
}

