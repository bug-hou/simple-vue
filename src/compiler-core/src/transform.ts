import { NodeTypes } from "./ast";
import { AST } from "./type/type";

export function transform(ast: AST, options?: any) {
  const context = createTransformContext(ast, options)
  // 遍历节点，使用深度优先算法
  traverseNode(ast, context)

  createRootCodegen(ast)

  ast.helpers = [...context.helpers.keys()]
}

function traverseNode(node: AST, context: any) {

  const nodeTransforms = context.nodeTransforms;
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    transform(node);
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper("toDisplayString")
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      const children = node.children;
      if (children) {
        console.log(node.type || node)
        children.forEach(item => {
          traverseNode(item, context)
        })
      }
      break;
    default:
      break;
  }

}

function createTransformContext(ast: AST, options: any) {
  const context = {
    ast,
    nodeTransforms: options?.nodeTransforms || [],
    helpers: new Map(),
    helper(k) {
      context.helpers.set(k, 1);
    }
  }
  return context;
}
function createRootCodegen(ast: AST) {
  ast.codegenNode = ast.children[0]
}

