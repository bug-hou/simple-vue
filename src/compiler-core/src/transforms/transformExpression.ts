import { NodeTypes } from "../ast";
import { AST } from "../type/type";

export function transformExpression(node: AST) {
  if (node.type === NodeTypes.INTERPOLATION) {
    node.content = processExpression(node.content);
  }
}

function processExpression(node: any) {
  node.content = `_ctx.${node.content}`;
  return node;
}