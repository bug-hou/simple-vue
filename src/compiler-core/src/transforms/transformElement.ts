import { NodeTypes } from "../ast";
import { CREATE_ELEMENT_VNODE } from "../runtimeHelpers";
import { AST } from "../type/type";

export function transformElement(node: AST, context: any) {
  if (node.type == NodeTypes.ELEMENT) {
    context.helper(CREATE_ELEMENT_VNODE);
    // 处理tag
    const vnodeTag = node.tag;
    let vnodeProps;

    let vnodeChildren = node.children[0];

    const vnodeElement = {
      type: NodeTypes.ELEMENT,
      tag: vnodeTag,
      props: vnodeProps,
      children: vnodeChildren
    }

    node.codegenNode = vnodeElement as any;
  }
}