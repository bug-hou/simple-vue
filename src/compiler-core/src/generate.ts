import { NodeTypes } from "./ast";
import { CREATE_ELEMENT_VNODE, helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers";
import { transformExpression } from "./transforms/transformExpression";
import { AST } from "./type/type";
import { isString } from "../../shared"

/* 
将ast转化为render函数
*/
export function generate(ast: AST) {

  const functionName = "render";
  const args = ["_ctx", "_cache", "$props", "$setup", "$data", "$options"];
  const signature = args.join(",");
  const context = createCodegenContext()
  let { push } = context;
  if (ast.helpers.length > 0) {
    genFunctionPreamble(ast, context);
  }

  push("return ")

  push(`function ${functionName}(${signature}){`)

  push("return ");

  genNode(ast.codegenNode, context);

  push("}")
  return {
    code: context.code
  }
}

function genNode(node: AST, context: any) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context)
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genSimpleExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;
  }
}


function createCodegenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper: helperMapName
  }
  return context;
}

function genFunctionPreamble(ast: AST, context) {
  const { push } = context;
  const VueBinging = "vue"
  const aliasHelper = s => `${helperMapName[s]} as _${helperMapName[s]}`
  push(`import { ${ast.helpers.map(aliasHelper).join(", ")} } from ${VueBinging}\r\n`)
}
function genText(node: AST, context) {
  const { push } = context;
  push(`'${node.content}'`)
}

function genInterpolation(node: AST, context: any) {
  const { push, helper } = context;
  push(`${helper[TO_DISPLAY_STRING]}(`)
  transformExpression(node)
  genNode(node.content as any, context);
  push(`)`)
}

function genSimpleExpression(node: AST, context: any) {
  const { push } = context;
  push(`${node.content}`)
}

function genElement(node: AST, context) {
  const { push, helper } = context;
  const { tag, children } = node;
  push(`${helper[CREATE_ELEMENT_VNODE]}("${tag}"),null,`)
  genNode(children as any, context);
  push(")")
  // if (children) {
  //   children.forEach(item => {
  //     genNode(item, context);
  //   })
  // }
}

function genCompoundExpression(node: AST, context) {
  const { push } = context;
  const children = node.children;
  if (children) {
    children.forEach(child => {
      if (isString(child)) {
        push(child)
      } else {
        genNode(child, context);
      }
    })
  }
}

