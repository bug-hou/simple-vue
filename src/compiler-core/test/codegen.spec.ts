import { baseParse } from "../src/parse"
import { generate } from "../src/generate"
import { transform } from "../src/transform";
import { transformElement } from "../src/transforms/transformElement";
import { transformText } from "../src/transforms/transformText";
import { transformExpression } from "../src/transforms/transformExpression";

describe("codegen", () => {
  it('string', () => {
    const ast = baseParse("hi");

    transform(ast)

    const { code } = generate(ast);

    expect(code).toMatchSnapshot()
  })
  it('interpolation', () => {
    const ast = baseParse("{{message}}");

    transform(ast)

    const { code } = generate(ast);

    expect(code).toMatchSnapshot()
  })
  it('element', () => {
    const ast = baseParse("<div>hi,{{message}}</div>");

    transform(ast, {
      nodeTransforms: [transformElement, transformText, transformExpression],
    })

    const { code } = generate(ast);

    expect(code).toMatchSnapshot()
  })
})

