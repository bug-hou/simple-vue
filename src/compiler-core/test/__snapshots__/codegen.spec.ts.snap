// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"import { _createElementVnode as __createElementVnode } from vue
return function render(_ctx,_cache,$props,$setup,$data,$options){return _createElementVnode(\\"div\\"),null,'hi,')}"
`;

exports[`codegen interpolation 1`] = `
"import { _toDisplayString as __toDisplayString } from vue
return function render(_ctx,_cache,$props,$setup,$data,$options){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `"return function render(_ctx,_cache,$props,$setup,$data,$options){return 'hi'}"`;
